### PR TITLE
Prov IP discovery via side-car agent

### PR DESCRIPTION
**Pull request by abays**

### CODE REVIEW

**Comment by abays** [comment]
I would prefer this one over the draft (init container) version I created yesterday.  Please let me know what you guys think.

File: controllers/provisionserver_controller.go [hunk 1]
```diff
@@ -130,37 +137,46 @@ func (r *ProvisionServerReconciler) Reconcile(req ctrl.Request) (ctrl.Result, er
...
-	instance.Status.LocalImageURL = r.getLocalImageURL(podIP, instance)
-	err = r.Client.Status().Update(context.TODO(), instance)
-
-	if err != nil {
-		r.Log.Error(err, "Failed to update CR status %v")
-		return ctrl.Result{}, err
+	// Provision IP Discovery Agent sets status' ProvisionIP
+	if instance.Status.ProvisionIP != "" {
+		urlPieces := strings.Split(instance.Status.LocalImageURL, "/")
+		if instance.Status.LocalImageURL == "" || len(urlPieces) <= 1 || (len(urlPieces) >= 3 && urlPieces[2] != instance.Status.LocalImageURL) {
```
Comment (stuggi) [inline controllers/provisionserver_controller.go:170]: could you comment here in which cases we update the status, it is a bit hard to follow from this if

**Review by stuggi — CHANGES_REQUESTED** [review]

File: controllers/provisionserver_controller.go (hunk 1 above)
Comment (abays) [reply controllers/provisionserver_controller.go:170]: Sure, my apologies.  I'll clean up the code (hopefully) by using the `net` package, and then add appropriate comments.

File: controllers/provisionserver_controller.go [hunk 2]
```diff
@@ -130,37 +137,46 @@ func (r *ProvisionServerReconciler) Reconcile(req ctrl.Request) (ctrl.Result, er
...
+			// Update status with LocalImageURL, given ProvisionIP status value
+			instance.Status.LocalImageURL = r.getLocalImageURL(instance)
+			err = r.Client.Status().Update(context.TODO(), instance)
+
+			if err != nil {
+				r.Log.Error(err, "Failed to update CR status %v")
+				return ctrl.Result{}, err
+			}
+		}
 	}
```
Comment (stuggi) [inline controllers/provisionserver_controller.go:192]: Thanks! This is now really easy to read :)

**Review by stuggi — APPROVED** [review]
